Could not find composite consistency error while creating & deleting realms in parallel

### Describe the bug

As mentionned in #60 , when running the `CreateDeleteRealms` scenario with more than 1 concurrent user, consistency failures are raised by Keycloak.

Scenario is working fine with a single user.

### Version

18.0.0

### Expected behavior

No consistency error while creating and deleting realms in parallel.

### Actual behavior

The `CreateDeleteRealms` scenario exhibits failures:
```
>>>>>>>>>>>>>>>>>>>>>>>>>>
Request:
Get admin-cli token: KO status.find.in(200,201,202,203,204,205,206,207,208,209,304), found 500
=========================
Session:
Session(Create then Delete realms,15,Map(clientId -> client-0, email -> , gatling.http.cache.dns -> io.gatling.http.resolver.ShufflingNameResolver@2f0afa56, state -> cdde7c76-d4ac-1023-29d2-89a18fd10320, redirectUri -> http://localhost:8080/realms/realm-0/account, username -> user-0, adminUsername -> keycloak, lastName -> , firstName -> , gatling.http.ssl.sslContexts -> io.gatling.http.util.SslContexts@6a0871d3, wrongPasswordCount -> 0, adminPassword -> keycloak, clientSecret -> client-0-secret, requestEpoch -> 1652101620232, realm -> realm-0, keycloakServer -> http://localhost:8080, gatling.http.cookies -> CookieJar(Map()), password -> user-0-password),KO,List(),io.gatling.core.protocol.ProtocolComponentsRegistry$$Lambda$597/0x00000008006a2040@6541c141,io.netty.channel.nio.NioEventLoop@435ce306)
=========================
HTTP request:
POST http://localhost:8080/realms/master/protocol/openid-connect/token
headers=
accept: application/json
host: localhost:8080
content-type: application/x-www-form-urlencoded
content-length: 75
stringBody=FormUrlEncodedRequestBody{contentType='application/x-www-form-urlencoded', charset=UTF-8, content=grant_type=password&client_id=admin-cli&username=keycloak&password=keycloak}
=========================
HTTP response:
status=
500 Internal Server Error
headers= 
Referrer-Policy: no-referrer
X-Frame-Options: SAMEORIGIN
Strict-Transport-Security: max-age=31536000; includeSubDomains
Cache-Control: no-store
X-Content-Type-Options: nosniff
Set-Cookie: KEYCLOAK_LOCALE=; Version=1; Comment=Expiring cookie; Expires=Thu, 01-Jan-1970 00:00:10 GMT; Max-Age=0; Path=/realms/master/; HttpOnly
Set-Cookie: KC_RESTART=; Version=1; Expires=Thu, 01-Jan-1970 00:00:10 GMT; Max-Age=0; Path=/realms/master/; HttpOnly
Pragma: no-cache
X-XSS-Protection: 1; mode=block
Content-Type: application/json
content-length: 25

body=
{"error":"unknown_error"}
<<<<<<<<<<<<<<<<<<<<<<<<<
```

The following stack is exhibited in Keycloak's logs:
```
2022-05-09 13:07:00,314 ERROR [org.keycloak.services.error.KeycloakErrorHandler] (executor-thread-1) Uncaught server error: java.lang.IllegalStateException: Could not find composite in role admin: 8e8c75ab-eda9-4359-9331-0b638d13a221
	at org.keycloak.models.cache.infinispan.RoleAdapter.lambda$getCompositesStream$0(RoleAdapter.java:137)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.HashMap$KeySpliterator.forEachRemaining(HashMap.java:1621)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
	at org.keycloak.models.cache.infinispan.RoleAdapter.getCompositesStream(RoleAdapter.java:140)
	at org.keycloak.models.utils.RoleUtils.expandCompositeRolesStream(RoleUtils.java:179)
	at org.keycloak.models.utils.RoleUtils.lambda$expandCompositeRoles$6(RoleUtils.java:201)
	at java.base/java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:271)
	at java.base/java.util.HashMap$KeySpliterator.forEachRemaining(HashMap.java:1621)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
	at org.keycloak.models.utils.RoleUtils.expandCompositeRoles(RoleUtils.java:202)
	at org.keycloak.models.utils.RoleUtils.getDeepUserRoleMappings(RoleUtils.java:223)
	at org.keycloak.protocol.oidc.TokenManager.getAccess(TokenManager.java:591)
	at org.keycloak.services.util.DefaultClientSessionContext.loadRoles(DefaultClientSessionContext.java:273)
	at org.keycloak.services.util.DefaultClientSessionContext.getRolesStream(DefaultClientSessionContext.java:141)
	at org.keycloak.utils.RoleResolveUtil.getAndCacheResolvedRoles(RoleResolveUtil.java:105)
	at org.keycloak.utils.RoleResolveUtil.getAllResolvedClientRoles(RoleResolveUtil.java:95)
	at org.keycloak.protocol.oidc.mappers.AudienceResolveProtocolMapper.transformAccessToken(AudienceResolveProtocolMapper.java:84)
	at org.keycloak.protocol.oidc.TokenManager.lambda$transformAccessToken$4(TokenManager.java:745)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at java.base/java.util.stream.SortedOps$RefSortingSink.end(SortedOps.java:395)
	at java.base/java.util.stream.Sink$ChainedReference.end(Sink.java:258)
	at java.base/java.util.stream.Sink$ChainedReference.end(Sink.java:258)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:485)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
	at org.keycloak.protocol.oidc.TokenManager.transformAccessToken(TokenManager.java:744)
	at org.keycloak.protocol.oidc.TokenManager.createClientAccessToken(TokenManager.java:534)
	at org.keycloak.protocol.oidc.TokenManager$AccessTokenResponseBuilder.generateAccessToken(TokenManager.java:1038)
	at org.keycloak.protocol.oidc.endpoints.TokenEndpoint.resourceOwnerPasswordCredentialsGrant(TokenEndpoint.java:627)
	at org.keycloak.protocol.oidc.endpoints.TokenEndpoint.processGrantRequest(TokenEndpoint.java:200)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:170)
	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:130)
	at org.jboss.resteasy.core.ResourceMethodInvoker.internalInvokeOnTarget(ResourceMethodInvoker.java:660)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTargetAfterFilter(ResourceMethodInvoker.java:524)
	at org.jboss.resteasy.core.ResourceMethodInvoker.lambda$invokeOnTarget$2(ResourceMethodInvoker.java:474)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTarget(ResourceMethodInvoker.java:476)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:434)
	at org.jboss.resteasy.core.ResourceLocatorInvoker.invokeOnTargetObject(ResourceLocatorInvoker.java:192)
	at org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:152)
	at org.jboss.resteasy.core.ResourceLocatorInvoker.invokeOnTargetObject(ResourceLocatorInvoker.java:183)
	at org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:141)
	at org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:32)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:492)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:261)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:161)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
	at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:164)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:247)
	at io.quarkus.resteasy.runtime.standalone.RequestDispatcher.service(RequestDispatcher.java:73)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatch(VertxRequestHandler.java:151)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.handle(VertxRequestHandler.java:82)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.handle(VertxRequestHandler.java:42)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1212)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:163)
	at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:141)
	at io.quarkus.vertx.http.runtime.StaticResourcesRecorder$2.handle(StaticResourcesRecorder.java:67)
	at io.quarkus.vertx.http.runtime.StaticResourcesRecorder$2.handle(StaticResourcesRecorder.java:55)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1212)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:163)
	at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:141)
	at io.quarkus.vertx.http.runtime.VertxHttpRecorder$5.handle(VertxHttpRecorder.java:380)
	at io.quarkus.vertx.http.runtime.VertxHttpRecorder$5.handle(VertxHttpRecorder.java:358)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1212)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:163)
	at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:141)
	at org.keycloak.quarkus.runtime.integration.web.QuarkusRequestFilter.lambda$createBlockingHandler$1(QuarkusRequestFilter.java:71)
	at io.vertx.core.impl.ContextImpl.lambda$null$0(ContextImpl.java:159)
	at io.vertx.core.impl.AbstractContext.dispatch(AbstractContext.java:100)
	at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$1(ContextImpl.java:157)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$13.runWith(VertxCoreRecorder.java:543)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

### How to Reproduce?

Setup Keycloak 18.0.0 using Postgres 14.2, then run the `CreateDeleteRealms` scenario with:
```
./kcb.sh --scenario=keycloak.scenario.admin.CreateDeleteRealms \
  --server-url=http://localhost:8080 \
  --concurrent-users=5 \
  --measurement=30 \
  --log-http-on-failure \
  --admin-username=<root admin username> \
  --admin-password=<root admin password>
```

### Anything else?

A possible problem is the transaction isolation level, which is `read committed` by default using Postgres.

Forcing the agroal connection pool to use a higher isolation level can be done using the following Java property:
`-Dquarkus.datasource.jdbc.transaction-isolation-level=repeatable-read`

Using this property ensures that the isolation level was properly set to `repeatable read` (checked by enabling JDBC calls logging using log4jdbc - not included in Keycloak current codebase).

The problem still occurs.

Note: Postgres does not allow phantom reads with the `repeatable read` transaction isolation, as described in https://www.postgresql.org/docs/current/transaction-iso.html :


Isolation Level | Dirty Read | Nonrepeatable Read | Phantom Read | Serialization Anomaly
-- | -- | -- | -- | --
Read uncommitted | Allowed, but not in PG | Possible | Possible | Possible
Read committed | Not possible | Possible | Possible | Possible
Repeatable read | Not possible | Not possible | Allowed, but not in PG | Possible
Serializable | Not possible | Not possible | Not possible | Not possible


Tracing transaction boundaries against issued SQL statements shows that all statements are executed within a unique transaction, as expected - so a non repeatable read problem is not supposed to happen (unless a bug in Postgres was introduced lately, but I highly doubt it).

> I suspect the problem occurs between step 2 and 3: In step 2, the realm was still there, and and therefore its composite role ID. Between the steps 2 and 3 the changes are committed, and when reading the roles by their composite ID, they are now gone.

That's what I supposed too, but this requires a transaction boundary or isolation problem - which I could not find.

> This doesn't violate the repeatable read as those composite roles haven't been read before strep 3.

I don't think this is correct: at least in Postgres and unless I'm very wrong (please tell me, and I'll get back to my books ;-) ), the repeatable read transaction running the roles collection logic *cannot see* any deletion/update/insert that has been committed after the transaction started.

Here is a simple case to illustrate:

Setup the tables and initial data:
```
BEGIN TRANSACTION ISOLATION LEVEL REPEATABLE READ;
    DROP TABLE IF EXISTS parentchild;
    DROP TABLE IF EXISTS entity;

    CREATE TABLE entity (
        id VARCHAR(32) NOT NULL,
        PRIMARY KEY(id)
    );
    
    CREATE TABLE parentchild (
        parentid VARCHAR(32) NOT NULL,
        childid VARCHAR(32) NOT NULL,
        PRIMARY KEY(parentid, childid),
        CONSTRAINT fk_parent FOREIGN KEY(parentid) REFERENCES entity(id),
        CONSTRAINT fk_child FOREIGN KEY(parentid) REFERENCES entity(id)
    );
COMMIT;


BEGIN TRANSACTION ISOLATION LEVEL REPEATABLE READ;
    INSERT INTO entity VALUES ( '001' );
    INSERT INTO entity VALUES ( '002' );
    INSERT INTO entity VALUES ( '003' );

    INSERT INTO parentchild VALUES ( '001' , '002');
    INSERT INTO parentchild VALUES ( '001' , '003');
COMMIT;
```

=> Transaction 1:
```
-- Txn 1
BEGIN TRANSACTION ISOLATION LEVEL REPEATABLE READ;
    DELETE FROM parentchild WHERE childid = '003';
    DELETE FROM entity WHERE id = '003';
```

Output:
```
BEGIN
DELETE 1
DELETE 1
```

=> Transaction 2:
```
-- Txn 2
BEGIN TRANSACTION ISOLATION LEVEL REPEATABLE READ;
    SELECT * FROM entity WHERE id = '001';
    SELECT * FROM parentchild WHERE parentid = '001';
```

Output:
```
BEGIN
 id  
-----
 001
(1 row)

 parentid | childid 
----------+---------
 001      | 002
 001      | 003
(2 rows)
```

=> Back to Transaction 1:
```
COMMIT;
```

Output:
```
COMMIT
```

=> Back to Transaction 2:
```
    SELECT * FROM entity WHERE id = '002';
    SELECT * FROM entity WHERE id = '003';
COMMIT;
```

Output:
```
 id  
-----
 002
(1 row)

 id  
-----
 003
(1 row)

COMMIT;
```

As you can see, Transaction 2 still sees the `entity` records with ids `002` and `003`, despite the Transaction 1 having deleted (and committed) id `003` before it reads them.

Something in my reasoning / somewhere else looks wrong...

> Therefore I assume this anomaly could only be avoided by using the Serializable isolation level. Could you please give it a try?

Sure.
Another class of problem arises:
```
2022-05-09 15:17:55,672 ERROR [org.keycloak.services.error.KeycloakErrorHandler] (executor-thread-4) Uncaught server error: org.keycloak.models.ModelException: javax.persistence.OptimisticLockException: org.hibernate.exception.LockAcquisitionException: could not execute statement
	at org.keycloak.connections.jpa.PersistenceExceptionConverter.convert(PersistenceExceptionConverter.java:84)
	at org.keycloak.connections.jpa.PersistenceExceptionConverter.invoke(PersistenceExceptionConverter.java:62)
	at com.sun.proxy.$Proxy76.flush(Unknown Source)
	at org.keycloak.models.jpa.JpaRealmProvider.addRealmRole(JpaRealmProvider.java:227)
	at org.keycloak.storage.RoleStorageManager.addRealmRole(RoleStorageManager.java:122)
	at org.keycloak.models.cache.infinispan.RealmCacheSession.addRealmRole(RealmCacheSession.java:624)
	at org.keycloak.models.cache.infinispan.RealmCacheSession.addRealmRole(RealmCacheSession.java:619)
	at org.keycloak.models.jpa.RealmAdapter.addRole(RealmAdapter.java:839)
	at org.keycloak.models.utils.KeycloakModelUtils.setupAuthorizationServices(KeycloakModelUtils.java:736)
	at org.keycloak.services.managers.RealmManager.setupAuthorizationServices(RealmManager.java:710)
	at org.keycloak.services.managers.RealmManager.importRealm(RealmManager.java:595)
	at org.keycloak.services.managers.RealmManager.importRealm(RealmManager.java:495)
	at org.keycloak.services.resources.admin.RealmsAdminResource.importRealm(RealmsAdminResource.java:133)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:170)
	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:130)
	at org.jboss.resteasy.core.ResourceMethodInvoker.internalInvokeOnTarget(ResourceMethodInvoker.java:660)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTargetAfterFilter(ResourceMethodInvoker.java:524)
	at org.jboss.resteasy.core.ResourceMethodInvoker.lambda$invokeOnTarget$2(ResourceMethodInvoker.java:474)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTarget(ResourceMethodInvoker.java:476)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:434)
	at org.jboss.resteasy.core.ResourceLocatorInvoker.invokeOnTargetObject(ResourceLocatorInvoker.java:192)
	at org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:141)
	at org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:32)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:492)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:261)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:161)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
	at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:164)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:247)
	at io.quarkus.resteasy.runtime.standalone.RequestDispatcher.service(RequestDispatcher.java:73)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatch(VertxRequestHandler.java:151)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.handle(VertxRequestHandler.java:82)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.handle(VertxRequestHandler.java:42)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1212)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:163)
	at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:141)
	at io.quarkus.vertx.http.runtime.StaticResourcesRecorder$2.handle(StaticResourcesRecorder.java:67)
	at io.quarkus.vertx.http.runtime.StaticResourcesRecorder$2.handle(StaticResourcesRecorder.java:55)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1212)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:163)
	at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:141)
	at io.quarkus.vertx.http.runtime.VertxHttpRecorder$5.handle(VertxHttpRecorder.java:380)
	at io.quarkus.vertx.http.runtime.VertxHttpRecorder$5.handle(VertxHttpRecorder.java:358)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1212)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:163)
	at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:141)
	at org.keycloak.quarkus.runtime.integration.web.QuarkusRequestFilter.lambda$createBlockingHandler$1(QuarkusRequestFilter.java:71)
	at io.vertx.core.impl.ContextImpl.lambda$null$0(ContextImpl.java:159)
	at io.vertx.core.impl.AbstractContext.dispatch(AbstractContext.java:100)
	at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$1(ContextImpl.java:157)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$13.runWith(VertxCoreRecorder.java:543)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: javax.persistence.OptimisticLockException: org.hibernate.exception.LockAcquisitionException: could not execute statement
	at org.hibernate.internal.ExceptionConverterImpl.wrapLockException(ExceptionConverterImpl.java:277)
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:98)
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:181)
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:188)
	at org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1406)
	at org.hibernate.internal.SessionImpl.flush(SessionImpl.java:1389)
	at jdk.internal.reflect.GeneratedMethodAccessor207.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.keycloak.connections.jpa.PersistenceExceptionConverter.invoke(PersistenceExceptionConverter.java:60)
	... 60 more
Caused by: org.hibernate.exception.LockAcquisitionException: could not execute statement
	at org.hibernate.exception.internal.SQLStateConversionDelegate.convert(SQLStateConversionDelegate.java:120)
	at org.hibernate.exception.internal.StandardSQLExceptionConverter.convert(StandardSQLExceptionConverter.java:37)
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:113)
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:99)
	at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.executeUpdate(ResultSetReturnImpl.java:200)
	at org.hibernate.engine.jdbc.batch.internal.NonBatchingBatch.addToBatch(NonBatchingBatch.java:46)
	at org.hibernate.persister.entity.AbstractEntityPersister.insert(AbstractEntityPersister.java:3375)
	at org.hibernate.persister.entity.AbstractEntityPersister.insert(AbstractEntityPersister.java:3908)
	at org.hibernate.action.internal.EntityInsertAction.execute(EntityInsertAction.java:107)
	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:604)
	at org.hibernate.engine.spi.ActionQueue.lambda$executeActions$1(ActionQueue.java:478)
	at java.base/java.util.LinkedHashMap.forEach(LinkedHashMap.java:684)
	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:475)
	at org.hibernate.event.internal.AbstractFlushingEventListener.performExecutions(AbstractFlushingEventListener.java:344)
	at org.hibernate.event.internal.DefaultFlushEventListener.onFlush(DefaultFlushEventListener.java:40)
	at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:107)
	at org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1402)
	... 65 more
Caused by: org.postgresql.util.PSQLException: ERROR: could not serialize access due to read/write dependencies among transactions
  Detail: Reason code: Canceled on identification as a pivot, during conflict in checking.
  Hint: The transaction might succeed if retried.
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2675)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2365)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:355)
	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:490)
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:408)
	at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:166)
	at org.postgresql.jdbc.PgPreparedStatement.executeUpdate(PgPreparedStatement.java:134)
	at jdk.internal.reflect.GeneratedMethodAccessor195.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.postgresql.ds.PGPooledConnection$StatementHandler.invoke(PGPooledConnection.java:441)
	at com.sun.proxy.$Proxy84.executeUpdate(Unknown Source)
	at io.agroal.pool.wrapper.PreparedStatementWrapper.executeUpdate(PreparedStatementWrapper.java:88)
	at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.executeUpdate(ResultSetReturnImpl.java:197)
	... 77 more
```

This is an expected and documented behavior with the `serializable`: code should expect this to happen, rollback and retry (which is not done anyway, I think).

Not sure why the exception refers to optimistic locking, since I cannot recall anything related to it in the legacy store (I may be wrong): `javax.persistence.OptimisticLockException`

> Basically this could happen "any time" if a composite role is update. I wonder if Keycloak should in such cases more gracefully and skip this composite role it isn't contained in the store any more. This is a rare occurrence on the other hand it would hide an potential inconsistency in the store if we would ignore it all the time. Maybe issue a warning... WDYT? But let's first complete the analysis before rushing to any code change.

Irrespective of this investigation, the yet-to-come optimisation PR I did drops this consistency check (i.e. it does not check that all roles have been successfully loaded from the provided IDs), because listing roles that can't be found in the database is in any case not correct, and doing so also defeats the `Stream`-oriented approach  - but since this is an explicit check in the current codebase, I did left a TODO mark to discuss this exact question.

--- a/model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/RoleAdapter.java
+++ b/model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/RoleAdapter.java
@@ -130,14 +130,15 @@ public class RoleAdapter implements RoleModel {
 
         if (composites == null) {
             composites = new HashSet<>();
-            composites = cached.getComposites().stream()
-                    .map(id -> {
-                        RoleModel role = realm.getRoleById(id);
-                        if (role == null) {
-                            throw new IllegalStateException("Could not find composite in role " + getName() + ": " + id);
-                        }
-                        return role;
-                    }).collect(Collectors.toSet());
+            for (String id : cached.getComposites()) {
+                RoleModel role = realm.getRoleById(id);
+                if (role == null) {
+                    // chance that composite role was removed, so invalidate this entry and fallback to delegate
+                    getDelegateForUpdate();
+                    return updated.getCompositesStream();
+                }
+                composites.add(role);
+            }
         }
 
         return composites.stream();

